### ISSUE TITLE
rest arguments optimization regression with nested functions

### ISSUE THREAD

**Issue by mathieumg** [babel/babel#1815]
At first I thought this was due to https://github.com/babel/babel/issues/1796 but that is fixed and this issue is slightly different and still happens on 5.6.4.

It works correctly in the REPL, which seems to be at 5.5.1:http://babeljs.io/repl/#?experimental=true&evaluate=true&loose=false&spec=false&code=function%20a(...args)%20%7B%0A%20%20return%20function()%20%7B%0A%20%20%20%20function%20b()%20%7B%7D%0A%0A%20%20%20%20console.log(%22Shouldn't%20args%20be%20from%20a's%20scope%3F%22%2C%20args)%3B%0A%20%20%7D%3B%0A%7D%0A

The same code outputs the following in 5.6.4:

```js
"use strict";

function a() {
  return function () {
    function b() {}

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    console.log("Shouldn't args be from a's scope?", args);
  };
}
```

It works fine in both versions if the `function b() {}` declaration is removed. (I imagine you don't need me to post snippets of that version as well?) Note that it's a simplified version of the original problem which comes from https://github.com/react-bootstrap/react-bootstrap/commit/0d9a933d52871a44119012a1f3ddd0fdfdb4b3f0#diff-f55b29b292e6d6afb446c72ed6f099cfR80, to help with pinpointing the issue.

Let me know if you need additional info or if this the intended behavior per the spec, but that would seem odd.

**Comment by sebmck** [comment]
Looks like it's a result of the sibling function causing the deopt flag to be set to `false` instead of the previous value. Fixed as of 5.6.4.

**Comment by mathieumg** [comment]
5.6.5 :stuck_out_tongue: 

Thanks for the fix, that was quick!
